### PR TITLE
Optimized Compute Score Function

### PR DESCRIPTION
**Pull request by jimip85**
In this PR, 

I refactored the computeScore function in FirstProgrammingLanguageQuiz.tsx for improved performance and maintainability. It utilizes the reduce() to calculate quiz score and updates score state object using setScore hook. I also introduced a null-check before calling computeScore in useEffect hook. To ensures that quiz score is correctly calculated and updated. In addition, I relocated the code into the API folder for server-side usability. 

Also, I have adopted the computeScore functionality to work in conjunction with the maxLangScore function in the LangResult page to dynamically render images.

Upcoming tasks: 
(a) implement unit tests to validate functionality.

### CODE REVIEW

File: components/resources/quizzes/shared/SkillSelections.tsx
Comment (LuckyHariharan) [inline components/resources/quizzes/shared/SkillSelections.tsx:16]: make sure to specify type before merging

File: pages/api/studentPortal/quizzes/scoringLogicFPL.ts
Comment (LuckyHariharan) [inline pages/api/studentPortal/quizzes/scoringLogicFPL.ts:3]: good use of reduce, makes sense to me to use this iterator. is it possible to simplify this and get away with just one reduce? not sure if its possible just curious

File: components/resources/quizzes/langQuiz/LangResults.tsx
Comment (LuckyHariharan) [inline components/resources/quizzes/langQuiz/LangResults.tsx:57]: great work covering the edgecases!

File: components/resources/quizzes/langQuiz/LangResults.tsx
Comment (vithushan19) [inline components/resources/quizzes/langQuiz/LangResults.tsx:4]: lets stop using the short form Lang, use the full word Language. It might feel verbose to you and faster to use the short form, but it's more readable to use more descriptive full words

File: components/resources/quizzes/langQuiz/LangResults.tsx
Comment (vithushan19) [inline components/resources/quizzes/langQuiz/LangResults.tsx:6]: Use the typescript Record  type and type aliases for this. 

```
type LanguageQuizResultsMap = Record<string, number>;
```

```
score: LanguageQuizResultsMap
```

https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-aliases
https://stackoverflow.com/questions/51936369/what-is-the-record-type

File: components/resources/quizzes/langQuiz/LangResults.tsx
Comment (vithushan19) [inline components/resources/quizzes/langQuiz/LangResults.tsx:9]: rename to LanguageResults

File: components/resources/quizzes/langQuiz/LangResults.tsx
Comment (vithushan19) [inline components/resources/quizzes/langQuiz/LangResults.tsx:10]:
```
const getPreferredLanguageForQuizResults = (score: LanguageQuizResultsMap) => { ... }
```

Comment (vithushan19) [reply components/resources/quizzes/langQuiz/LangResults.tsx:10]: write a unit test in this PR that tests this function with various inputs

Comment (vithushan19) [reply components/resources/quizzes/langQuiz/LangResults.tsx:10]: this function should but move into the api folder

File: components/resources/quizzes/langQuiz/LangResults.tsx
Comment (vithushan19) [inline components/resources/quizzes/langQuiz/LangResults.tsx:28]: language

File: components/resources/quizzes/langQuiz/LangResults.tsx
Comment (vithushan19) [inline components/resources/quizzes/langQuiz/LangResults.tsx:24]: types should be capitalized. Use a more descriptive name `FirstLanguageQuizResultData`

File: components/resources/quizzes/langQuiz/LangResults.tsx
Comment (vithushan19) [inline components/resources/quizzes/langQuiz/LangResults.tsx:32]: move to api folder

File: components/resources/quizzes/langQuiz/LangResults.tsx
Comment (vithushan19) [reply components/resources/quizzes/langQuiz/LangResults.tsx:57]: agreed!

File: components/resources/quizzes/langQuiz/LangResults.tsx
Comment (vithushan19) [inline components/resources/quizzes/langQuiz/LangResults.tsx:45]: the size of the image is view logic, but the rest of this object is business logic. Business logic should go in the api folder, but that means these sizes would also have to go into the api folder, which is not ideal.

Is it possible to use scale-75 for all images, and avoid having to store this information in the results object? Can you use a different image for python so we can use scale-75 everywhere

File: components/resources/quizzes/langQuiz/LangResults.tsx
Comment (vithushan19) [inline components/resources/quizzes/langQuiz/LangResults.tsx:68]: would be better if we could just hardcode scale-75 here or have a consistent tailwind classname for all quiz results

File: pages/api/studentPortal/quizzes/scoringLogicFPL.ts
Comment (vithushan19) [reply pages/api/studentPortal/quizzes/scoringLogicFPL.ts:3]: this is hard to read, ask chatgpt to rewrite it for you to make it more readable. Add unit tests for this function in this PR. 

If you still can't think of a cleaner way to write this, I'll try to help.

**Review by vithushan19 — CHANGES_REQUESTED** [review]
good first attempt. This PR is much more focused on the business logic so we should emphasize unit tests in this PR.

Also be careful editing shared components. When you edit a shared component, you need to take extra care that you're not breaking things for other people that might be using the same components. 

Also use more descriptive function and variable names. You use a lot of confusing technical jargon in your naming. Naming is one of the hardest problems in software development.

Here are some style guides that you should review:
https://google.github.io/styleguide/tsguide.html
https://ts.dev/style/#descriptive-names

Overall, good work. 
- Keep building out this feature, you are levelling up.
- Start looping in @kwiedenhaupt into your work. You'd benefit from more conversations with her.
- Take the time to read the resources I share. I don't give you all these resources upfront, because it can be dry and boring and less effective to be reading documentation all the time. Keep the focus on building and talking to people. But when I shine a light on a knowledge gap, take the time to go through these resources as if you were in school and had assigned readings or a textbook.

File: components/resources/quizzes/langQuiz/LangResults.tsx
Comment (jimip85) [reply components/resources/quizzes/langQuiz/LangResults.tsx:68]: okay - sounds good!

File: pages/api/studentPortal/quizzes/scoringLogicFPL.ts
Comment (jimip85) [reply pages/api/studentPortal/quizzes/scoringLogicFPL.ts:4]: `const score = { JavaScript: 0, "HTML/CSS": 0, Python: 0 };
  
  quizViewState.questions.map((question) => {
    question.options.map((option) => {
      if (option.isSelected && option.result) {
        score[option.result] += option.weight;
      }
    });
  });
  return score;`

Comment (jimip85) [reply pages/api/studentPortal/quizzes/scoringLogicFPL.ts:4]: Come up with a simplified solution!

File: components/resources/quizzes/langQuiz/LangResults.tsx
Comment (jimip85) [reply components/resources/quizzes/langQuiz/LangResults.tsx]: Done!

File: pages/api/studentPortal/quizzes/scoringLogicFPL.ts
Comment (LuckyHariharan) [reply pages/api/studentPortal/quizzes/scoringLogicFPL.ts:5]: excellent work Jimi! Super clean code, and makes my job implementing similar career-quiz results logic so much easier!

**Review by LuckyHariharan — COMMENTED** [review]
Excellent PR Jimi! Your score logic is clean and I am confident it will run well. Let me know if you want to chat about unit tests for the next PR to be 200% confident; this PR is already pretty big at +100 lines. Definitely ready for review @vithushan19

File: pages/api/studentPortal/quizzes/scoringLogicFPL.ts
Comment (vithushan19) [inline pages/api/studentPortal/quizzes/scoringLogicFPL.ts:9]: use a foreach instead of a mp in this line and the line above. Using map implies that you're creating a new transformed array. It's weird to see a map being used and then you don't do anything with the array. 

What you're looking to do is iterate through the arrays and add to the score variable. Since you're `mutating/editing` the score variable you should use `foreach`

File: pages/api/studentPortal/quizzes/scoringLogicFPL.ts
Comment (vithushan19) [inline pages/api/studentPortal/quizzes/scoringLogicFPL.ts:3]: looks like you have multiple files in the quizzes folder that are related to your quiz, I would create a subfolder for your quiz and move your files into that folder

File: pages/api/studentPortal/quizzes/scoringLogicFPL.ts
Comment (vithushan19) [reply pages/api/studentPortal/quizzes/scoringLogicFPL.ts:5]: you're new solution looks good, lets add a unit test for this function

File: pages/api/studentPortal/quizzes/scoringLogicFPL.ts
Comment (vithushan19) [inline pages/api/studentPortal/quizzes/scoringLogicFPL.ts:19]: add a unit test for this function as well

**Review by vithushan19 — COMMENTED** [review]
this looks great to me nice work. really impressive to see how quickly you were able to be coached to get this PR into this state. Code is readable, and the PR is small and easy to review.

Consider how you're earlier PRs required a lot of back and forth iterations. This one seems to have gone much smoother.

The only thing blocking the PR from getting approved is adding a unit test for your api functions (computeScore, getPreferredLanguageForQuizResults). It's best practice to include unit tests in the same PR, and not to do them separately.

File: pages/api/studentPortal/quizzes/scoringLogicFPL/computeScore.ts
Comment (vithushan19) [inline pages/api/studentPortal/quizzes/scoringLogicFPL/computeScore.ts:2]: rename the folder `scoringLogicFPL` to `firstProgrammingLanguage`

**Review by vithushan19 — COMMENTED** [review]
almost there, just a few more tests cases, and more consistent naming for your feature.

Try to use firstProgrammingLanguage everywhere to be consistent

File: pages/api/studentPortal/quizzes/scoringLogicFPL/computeScore.test.ts
Comment (jimip85) [reply pages/api/studentPortal/quizzes/scoringLogicFPL/computeScore.test.ts:76]: Hey Vithushan - since this function solely summates weight values - aren't some of these tests redundant?? 

I feel that the preferredLanguageQuiz component would benefit from these edge-cases!

What do you think?

Comment (vithushan19) [reply pages/api/studentPortal/quizzes/scoringLogicFPL/computeScore.test.ts:76]: @jimip85 makes sense, if you feel like the tests are redundant, you don't have to add them. But I'm in the camp of better being safe than sorry. 

I believe that we may need to adapt your scoring logic in the future to do more than just calculate weights, and if the complexity of score calculations grow and change, we'd want to make sure these tests always pass. 

Having more complete coverage will help protect your feature as other people work on it, so I'm generally ok to add more test coverage. If you truly feel that they're pointless then feel free to push back.

I'm good with keeping the tests here

File: pages/api/studentPortal/quizzes/scoringLogicFPL.ts
Comment (jimip85) [reply pages/api/studentPortal/quizzes/scoringLogicFPL.ts]: Done!

**Review by vithushan19 — APPROVED** [review]
nice work!

**Pull request merged**
